### PR TITLE
Add EfficientNet example using automatic augmentations with DALI

### PR DESCRIPTION
**Pull request by klecki**
## Description:
This example ports the [EfficientNet](https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Classification/ConvNets/efficientnet/README.md) sample from DeepLearningExamples repository.

The example is limited to efficientnet-b0 variant for simplicity. DALI pipeline is updated to use `fn` API and to use new automatic augmentations: adding options to select both AutoAugment and TrivialAugment.

The main.py is adjusted so the defaults are suitable for EfficientNet training immediately (previously they were the defaults for RN50 training) and launch.py is no longer needed - the original example was started via launch.py, that looked up default values for specific network in an .yml config and passed them to the main.py. This way we can use main.py directly without the layers of intermediate scripts. 

The benchmarks from readme are used to implement the L3 test.

The automatic augmentations come from: https://github.com/NVIDIA/DALI/pull/4648. It can already be reviewed as the API is basically one-line invocation within the pipeline definition.

### Affected modules and functionalities:
Docs/examples PR with L3 test.

### Key points relevant for the review:

❗ Please check if the defaults in main.py match the ones in https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Classification/ConvNets/configs.yml for DGX-1V, efficientnet-b0.

## How to review this PR
I suggest checking out the code and running diff tool like `meld` to compare directories:

 * `docs/examples/use_cases/pytorch/efficientnet/` from this PR 
 * `PyTorch/Classification/ConvNets/` from [DeepLearningExamples](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Classification/ConvNets)

that way you can see that most of this PR are just files copied over.

### Individual commits description:
I tried my best to split the PR into commits that are easier to review. Those are the steps (and specific commits):
1. Copy the contents of DeepLearningExamples `PyTorch/Classification/ConvNets/` 
2. Remove as many files as possible from the resulting directory.
3. Remove the usage of layouts other than efficientnet-b0
4. Propagating the defaults from https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Classification/ConvNets/configs.yml to main.py
5. Add new DALI pipeline, readme, new arguments <- most changes here
    * new pipeline is defined in dali.py
    * old is removed from dataloaders.py
    * main.py gets some argument changes
    * readme.rst is introduced.
7. Add new L3 test  + some small fixes for pipeline.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [x] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [x] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: 3194

### CODE REVIEW

**Comment by klecki** [comment]
Todo, apply in the test:
```
[PASS] bench_report_synthetic.json above threshold: 11213.31192785647 >= 10800                                                                                                                                                                                                                [PASS] bench_report_dali.json above threshold: 9440.822217370116 >= 6000                                                                                                                                                                                                                      [PASS] bench_report_dali_aa.json above threshold: 9535.54048843674 >= 9000                                                                                                                                                                                                                    [PASS] bench_report_dali_ta.json above threshold: 9485.628250563821 >= 9000                                                                                                                                                                                                                   [PASS] bench_report_pytorch.json above threshold: 7501.932983753756 >= 7200                                                                                                                                                                                                                   [PASS] bench_report_pytorch_aa.json above threshold: 7329.357160196307 >= 7200            
```
+ reduce the number of iterations by half, possibly diable some variants.

**Comment by jantonguirao** [comment]
Can you make this PR to have a first commit with a copy of what's in DeepLearningExamples so that I can see what you actually changed?

**Comment by klecki** [comment]
> Can you make this PR to have a first commit with a copy of what's in DeepLearningExamples so that I can see what you actually changed?

It is already done this way, look at the PR description for details about individual commits.

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    # evaluated exactly once during build - we either include automatic augmentations or not.
+    if automatic_augmentation == "autoaugment":
+        shapes = fn.peek_image_shape(jpegs)
+        output = auto_augment.auto_augment_image_net(images, shapes)
+    elif automatic_augmentation == "trivialaugment":
+        output = trivial_augment.trivial_augment_wide(images)
+    else:
+        output = images
+
+    output = fn.crop_mirror_normalize(output, dtype=types.FLOAT, output_layout=types.NCHW,
```
Comment (jantonguirao) [inline docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:62]:
```suggestion
    output = fn.crop_mirror_normalize(output, dtype=types.FLOAT, output_layout="CHW",
```
`types.NCHW` was deprecated years ago

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+@pipeline_def
+def validation_pipe(data_dir, interpolation, image_size, image_crop, rank=0, world_size=1):
+    jpegs, label = fn.readers.file(file_root=data_dir, shard_id=rank, num_shards=world_size,
+                                   random_shuffle=False, pad_last_batch=True)
+
+    images = fn.decoders.image(jpegs, device="mixed", output_type=types.RGB)
+
+    images = fn.resize(images, resize_shorter=image_size, interp_type=interpolation,
+                       antialias=False)
```
Comment (jantonguirao) [inline docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:78]: do you really want to disable antialiasing? Just asking

**Review by jantonguirao — APPROVED** [review]
LGTM, minor comments only

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dataloaders.py [hunk 3]
```diff
@@ -40,17 +38,19 @@
...
+
+    DATA_BACKEND_CHOICES.append("dali")
+except ImportError as e:
     print(
         "Please install DALI from https://www.github.com/NVIDIA/DALI to run this example."
     )
 
+# TODO(klecki): Move it back again
+import torchvision.datasets as datasets
+import torchvision.transforms as transforms
```
Comment (szalpal) [inline docs/examples/use_cases/pytorch/efficientnet/image_classification/dataloaders.py:53]: I don't know if this TODO is a leftover or something that should stay. However, if it should stay, I believe it should describe better what should be done ;)

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    if dali_device == "gpu":
+        decoder_device = "mixed"
+        rrc_device = "gpu"
+    else:
+        decoder_device = "cpu"
+        rrc_device = "cpu"
+
+    images = fn.decoders.image(jpegs, device=decoder_device, output_type=types.RGB,
+                               device_memory_padding=211025920, host_memory_padding=140544512)
```
Comment (szalpal) [inline docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:39]: 🤯

**Review by szalpal — COMMENTED** [review]
Put some comments in the readme. I still have the `sh` file left to review

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py (hunk 2 above)
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:78]: I took it from the original pipeline.

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+      * Scale from 8% to 100%
+      * Aspect ratio from 3/4 to 4/3
+
+    * Random horizontal flip
+    * [Optional: AutoAugment or TrivialAugment]
+    * Normalization
+
+* For inference:
+
+  * Scale to target image size + 32
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:38]: It's just 224 + 32 - the definition was taken from the original model as was (partially) this description. I will reword it a bit.

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py (hunk 1 above)
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:62]: I just realized, that we should have NHWC as a parameter here, and I wonder how it even got a good training result.

Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:62]: So we are doing a double transposition unnecessarily, I wonder if and how much faster we can get without it.

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 6]
```diff
@@ -0,0 +1,172 @@
+EfficientNet for PyTorch with DALI and AutoAugment
+==================================================
+
+This example shows how DALI's implementation of automatic augmentations - most notably  `AutoAugment <https://arxiv.org/abs/1805.09501>`_ and `TrivialAugment <https://arxiv.org/abs/2103.10158>`_ - can be used in training. It shows the training of EfficientNet, an image classification model first described in  `EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks <https://arxiv.org/abs/1905.11946>`_.
+
+The code is based on `NVIDIA Deep Learning Examples <https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch/Classification/ConvNets/efficientnet>`_ - it has been extended with DALI pipeline supporting automatic augmentations, which can be found in :fileref:`here <docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py>`.
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:6]: It does in the sphinx, I took it from SSD example. I will try to please both Sphinx and GH, but I don't know if it is possible.

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py (hunk 1 above)
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py]: We can get around the transposition, I am not sure if it gives any benefit, the benchmarks give me a bit more samples/s but that might be just noise.

Either way, I use "CHW" and "HWC" now and produce the memory in target layout for NHWC case.

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/models/__init__.py [hunk 7]
```diff
@@ -13,11 +13,4 @@
...
 from .entrypoints import nvidia_convnets_processing_utils, nvidia_efficientnet
-from .efficientnet import (
-    efficientnet_b0,
-    efficientnet_b4,
-    efficientnet_widese_b0,
-    efficientnet_widese_b4,
-    efficientnet_quant_b0,
-    efficientnet_quant_b4,
-)
+from .efficientnet import efficientnet_b0
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/models/__init__.py:16]: done

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+
+Differences to the Deep Learning Examples configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* The default values of the parameters were adjusted to values used in EfficientNet training.
+* ``--data-backend`` parameter was changed to accept ``dali``, ``pytorch``, or ``synthetic``. It is set to ``dali`` by default.
+* ``--dali-device`` was added to control placement of some of DALI operators.
+* ``--augmentation`` was replaced with ``--automatic-augmentation``, now supporting ``disabled``, ``autoaugment``, and ``trivialaugment`` values.
+* ``--workers`` defaults were halved to accommodate DALI. The value is automatically doubled when ``pytorch`` data loader is used.
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:16]: I added a bit about fitting both loaders with good default, but I am not sure if I really want to dive deep into how it works.

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+* the model is restricted to EfficientNet-B0 architecture.
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:17]: done

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+* For inference:
+
+  * Scale to target image size + 32
+  * Center crop to target image size
+  * Normalization
+
+
+
+Requirements
+^^^^^^^^^^^^
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:45]: I changed it to Setup

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+Training
+--------
+
+To run training on 1GPU, use the ``main.py`` entry point:
+
+* For FP32: ``python ./main.py $PATH_TO_IMAGENET``
+* For AMP: ``python ./main.py --amp --static-loss-scale 128 $PATH_TO_IMAGENET``
+
+You may need to adjust ``--batch-size`` parameter for your machine.
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:85]: done

File: docs/examples/use_cases/pytorch/efficientnet/readme.rst [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+You can change the data loader and automatic augmentation scheme that are used by adding:
+
+* ``--data-backend``: dali | pytorch | synthetic,
+* ``--automatic-augmentation``: disabled | autoaugment | trivialaugment (the last one only for DALI),
+* ``--dali-device``: cpu | gpu (only for DALI).
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/readme.rst:91]: done

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py (hunk 4 above)
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:41]: The allocation padding is taken from the original model, I added a comment.

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dataloaders.py (hunk 3 above)
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dataloaders.py:53]: It doesn't matter much, but it breaks in my local setup as torchvision clashes with my local build of DALI, so I need to import DALI first.

**Review by banasraf — APPROVED** [review]

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    # Make sure that from this point we are processing on GPU regardless of dali_device parameter
+    images = images.gpu()
+
+    images = fn.flip(images, horizontal=rng)
+
+    # Based on the specification, apply the automatic augmentation policy. Note, that from the point
+    # of Pipeline definition, this `if` statement relies on static scalar parameter, so it is
+    # evaluated exactly once during build - we either include automatic augmentations or not.
+    if automatic_augmentation == "autoaugment":
+        shapes = fn.peek_image_shape(jpegs)
```
Comment (stiepan) [inline docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:58]: If the image sizes are uniform (and they are thanks to the `resize`) we can skip shapes and use just the absolute version. It uses mag_range=(0, 250) as in the original tf EfficientNet example by default so I guess it should be fine out-of-the-box.

**Review by szalpal — APPROVED** [review]

File: docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py (hunk 13 above)
Comment (klecki) [reply docs/examples/use_cases/pytorch/efficientnet/image_classification/dali.py:58]: I think it is better to show the more flexible version in this example if it doesn't cause perf issues. Let me check.
